File: src-cc/lin.cc
using namespace std;
#include<iostream>
#include<cmath>
#include<vector>
#include"lin.h"

template<typename T1> void swp(T1 &a, T1 &b)
{
    T1 tmp=a;
    a=b;
    b=tmp;
}

static double abs_max_vector(vector<double> &v, size_t from, size_t end)
{
    double vmax=fabs(v[from]);
    for(size_t i=from+1; i<end; i++){
	vmax=max(vmax, fabs(v[i]));
    }
    return vmax;
}

static void scale_vector(vector<double> &v, size_t from, size_t end, double factor)
{
    for(size_t i=from; i<end; i++){
	v[i]*=factor;
    }
}

static void scale_matrix_row(vector<vector<double> > &a, vector<double> &b)
{
    const size_t N=a.size();
    for(size_t i=0; i<N; i++){
	double factor=1.0/abs_max_vector(a[i],0,N);
	scale_vector(a[i],0,N,factor);
	b[i]*=factor;
    }
}

static void pivot(vector<vector<double> > & a, vector<double> &b, size_t i)
{
    double ajimax=fabs(a[i][i]);
    size_t maxj=i;
    const size_t N=a.size();

    for(size_t j=i+1; j<N; j++){
	double aji=fabs(a[j][i]);
	if(aji>ajimax){
	    ajimax=aji;
	    maxj=j;
	}
    }
    if(i!=maxj){
	swp(a[i],a[maxj]);
	swp(b[i],b[maxj]);
    }
}

static void update_lower_col(vector<vector<double> > &a, vector<double> &b,
	size_t i, size_t blockend)
{
    const size_t N=a.size();
    for(size_t i1=i; i1<blockend; i1++){
	pivot(a,b,i1);
	double factor=1.0/a[i1][i1];
	scale_vector(a[i1],i1+1,blockend,factor);
	b[i1]*=factor;

	for(size_t j=i1+1; j<N; j++){
	    factor=a[j][i1];
	    for(size_t k=i1+1; k<blockend; k++){
		a[j][k]-=factor*a[i1][k];
	    }
	    b[j]-=factor*b[i1];
	}
    }
}

static void update_upper_row(vector<vector<double> > &a, size_t i,
	size_t blockend)
{
    const size_t N=a.size();

    for(size_t i1=i; i1<blockend; i1++){
	scale_vector(a[i1],blockend,N,1.0/a[i1][i1]);
	for(size_t j=i1+1; j<blockend; j++){
	    for(size_t k=blockend; k<N; k++){
		a[j][k]-=a[j][i1]*a[i1][k];
	    }
	}
    }
}

static void forward_elimination(vector<vector<double> >&a, size_t i,
	size_t blockend)
{
    const size_t blocksize=48;
    const size_t N=a.size();

    for(size_t k=blockend; k<N; k+=blocksize){
	size_t kend=min(k+blocksize,N);
	for(size_t j=blockend; j<N; j++){
	    for(size_t l=i; l<blockend; l++){
		for(size_t k1=k; k1<kend; k1++){
		    a[j][k1]-=a[j][l]*a[l][k1];
		}
	    }
	}
    }
}

static void back_substitution(vector<vector<double> >&a, vector<double> &b)
{
    const size_t N=a.size();

    for(size_t i=N; i>0;){
	i--;
	for(size_t j=i+1; j<N; j++){
	    b[i]-=a[i][j]*b[j];
	}
    }
}

void solve(vector<vector<double> > &a, vector<double> &b)
{
    const size_t blocksize=48;
    const size_t N=a.size();

    /* scale */
    scale_matrix_row(a,b);
    for(size_t i=0; i<N; i+=blocksize){
	size_t blockend=min(i+blocksize, N);
	/* forward */
	update_lower_col(a,b,i,blockend);
	update_upper_row(a,i,blockend);
	forward_elimination(a,i,blockend);
    }
    /* back */
    back_substitution(a,b);
}

File: src-f/lin.f90
module lin
    implicit none
contains
subroutine swap(a, b)
    implicit none
    double precision, intent(inout):: a,b
    double precision :: tmp
    tmp=a
    a=b
    b=tmp
end subroutine

double precision function abs_max_array(x,maxi)
    implicit none
    double precision, intent(inout):: x(:)
    integer, intent(out):: maxi
    double precision:: xi
    integer i

    abs_max_array=abs(x(1))
    maxi=1
    do i=2, ubound(x,1)
        xi=abs(x(i))
        if(xi>abs_max_array)then
            abs_max_array=xi
            maxi=i
        endif
    enddo
end function

subroutine scale_matrix_row(a, b)
    implicit none
    double precision, intent(inout):: a(:,:), b(:)
    double precision :: factor
    integer :: i,N,dummy

    N=ubound(a,2)
    do i=1, ubound(a,1)
        factor=1.0d0/abs_max_array(a(i,1:N),dummy)
        a(i,1:N)=a(i,1:N)*factor
        b(i)=b(i)*factor
    enddo
end subroutine

subroutine pivot(a, b, i)
    implicit none
    double precision, intent(inout):: a(:,:), b(:)
    integer, intent(in) :: i
    double precision :: dummy
    integer :: N, j, maxj
    N=ubound(a,1)
    dummy=abs_max_array(a(i:N,i),maxj)
    maxj=maxj+i-1
    ! swap
    if(i/=maxj)then
        do j=i, ubound(a,2)
            call swap(a(i,j), a(maxj,j))
        enddo
        call swap(b(i), b(maxj))
    endif
end subroutine

subroutine update_lower_col(a, b, i, blend)
    implicit none
    double precision, intent(inout):: a(:,:), b(:)
    integer, intent(in):: i, blend
    double precision:: factor
    integer:: N,i1,j

    N=ubound(a,1)
    do i1=i, blend
        call pivot(a,b,i1)
        factor=1.0d0/a(i1,i1)
        a(i1,i1+1:blend)=a(i1,i1+1:blend)*factor
        b(i1)=b(i1)*factor

        do j=i1+1, blend
            a(i1+1:N,j)=a(i1+1:N,j)-a(i1+1:N,i1)*a(i1,j)
        enddo
        b(i1+1:N)=b(i1+1:N)-a(i1+1:N,i1)*b(i1)
    enddo
end subroutine

subroutine update_upper_row(a, i, blend)
    implicit none
    double precision,intent(inout):: a(:,:)
    integer, intent(in):: i, blend
    integer:: k,i1

    do k=blend+1, ubound(a,2)
        do i1=i, blend
            a(i1,k)=a(i1,k)/a(i1,i1)
            a(i1+1:blend,k)=a(i1+1:blend,k)-a(i1+1:blend,i1)*a(i1,k)
        enddo
    enddo
end subroutine

subroutine forward_elimination(a, i, blend)
    implicit none
    double precision,intent(inout):: a(:,:)
    integer, intent(in):: i, blend
    integer N,j,k,l,blsize
    integer jend

    blsize=48
    N=ubound(a,1)
    do j=blend+1, N, blsize
        jend=min(j+blsize-1,N)
        do k=blend+1, N
            do l=i, blend
                a(j:jend,k)=a(j:jend,k)-a(j:jend,l)*a(l,k)
            enddo
        enddo
    enddo
end subroutine

subroutine back_substitution(a, b)
    implicit none
    double precision, intent(in):: a(:,:)
    double precision, intent(inout):: b(:)
    integer:: i
    do i=ubound(b,1), 1, -1
        b(1:i-1)=b(1:i-1)-a(1:i-1,i)*b(i)
    enddo
end subroutine

subroutine solve(a, b)
    implicit none
    double precision, intent(inout):: a(:,:), b(:)
    integer :: i,N,blocksize, blend
    blocksize=48
    N=ubound(a,1)
    ! scale
    call scale_matrix_row(a,b)
    do i=1, N, blocksize
        blend=min(i+blocksize-1,N)
        ! forward
        call update_lower_col(a,b,i,blend)
        call update_upper_row(a,i,blend)
        call forward_elimination(a,i,blend)
    enddo
    ! backward
    call back_substitution(a,b)
end subroutine
end module
